src/sentry/api/authentication.py

```diff
@@ -306,10 +308,81 @@ def authenticate(self, request: Request):
         return self.transform_auth(user_id, None)
 
 
+class TokenStrLookupRequired(Exception):
+    """
+    Used in combination with `apitoken.use-and-update-hash-rate` option.
+
+    If raised, calling code should peform API token lookups based on its
+    plaintext value and not its hashed value.
+    """
+
+    pass
+
+
 @AuthenticationSiloLimit(SiloMode.REGION, SiloMode.CONTROL)
 class UserAuthTokenAuthentication(StandardAuthentication):
     token_name = b"bearer"
 
+    def _find_or_update_token_by_hash(self, token_str: str) -> ApiToken | ApiTokenReplica:
+        """
+        Find token by hash or update token's hash value if only found via plaintext.
+
+        1. Hash provided plaintext token.
+        2. Perform lookup based on hashed value.
+        3. If found, return the token.
+        4. If not found, search for the token based on its plaintext value.
+        5. If found, update the token's hashed value and return the token.
+        6. If not found via hash or plaintext value, raise AuthenticationFailed
+
+        Returns `ApiTokenReplica` if running in REGION silo or
+        `ApiToken` if running in CONTROL silo.
+        """
+
+        hashed_token = hashlib.sha256(token_str.encode()).hexdigest()
+
+        rate = options.get("apitoken.use-and-update-hash-rate")
+        random_rate = random.random()
+
+        if SiloMode.get_current_mode() == SiloMode.REGION:
+            try:
+                if rate > random_rate:
```

Comment:
Let's store this in `use_hashed_token = rate > random_rate` since it's duplicated a bit and it make things a little more readable